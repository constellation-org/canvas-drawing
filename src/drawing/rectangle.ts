type RectangleToDraw = {
  x: number
  y: number
  width: number
  height: number
  backgroundColor?: string
  borderColor?: string
  borderWidth?: number
  opacity?: number
}

const DEFAULT_OPACITY = 1
const DEFAULT_BORDER_WIDTH = 4
const DEFAULT_BORDER_COLOR = '#000'

export const drawRectangle = (
  context: CanvasRenderingContext2D,
  args: RectangleToDraw
) => {
  context.save()

  context.beginPath()
  context.rect(args.x, args.y, args.width, args.height)

  context.globalAlpha = args.opacity ?? DEFAULT_OPACITY

  if (args.backgroundColor) {
    context.fillStyle = args.backgroundColor
    context.fill()
  }

  context.strokeStyle = args.borderColor ?? DEFAULT_BORDER_COLOR
  context.lineWidth = args.borderWidth ?? DEFAULT_BORDER_WIDTH
  context.stroke()

  context.restore()
}
