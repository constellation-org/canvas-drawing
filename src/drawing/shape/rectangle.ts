import { ShapeToDraw, drawShape } from './shape'

export const drawRectangle = (
  context: CanvasRenderingContext2D,
  args: ShapeToDraw
) => {
  const path = (context: CanvasRenderingContext2D, args: ShapeToDraw) => {
    context.beginPath()
    context.rect(args.x, args.y, args.width, args.height)
  }
  drawShape(context, path, args)
}
