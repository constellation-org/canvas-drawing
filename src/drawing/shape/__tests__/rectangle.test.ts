import { drawRectangle } from '../rectangle'

describe('[Rectangle]', () => {
  let canvas: HTMLCanvasElement
  let context: CanvasRenderingContext2D
  beforeEach(function () {
    canvas = document.createElement('canvas')
    canvas.setAttribute('width', "100")
    canvas.setAttribute('height', "100")
    context = canvas.getContext('2d') as CanvasRenderingContext2D
  })
  it('should draw square on top left corner', () => {
    drawRectangle(context, { x: 0, y: 0, width: 20, height: 20})
    expect(canvas.toDataURL('image/png')).toMatchSnapshot()
  })
  it('should draw square on bottom right corner', () => {
    drawRectangle(context, { x: 80, y: 80, width: 20, height: 20})
    expect(canvas.toDataURL('image/png')).toMatchSnapshot()
  })
  it('should draw rectangle in the middle', () => {
    drawRectangle(context, { x: 30, y: 40, width: 40, height: 20})
    expect(canvas.toDataURL('image/png')).toMatchSnapshot()
  })
})
