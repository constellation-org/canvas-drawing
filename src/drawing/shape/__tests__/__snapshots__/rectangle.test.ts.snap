// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Rectangle] should draw rectangle in the middle 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAAAXNSR0IArs4c6QAAAgpJREFUeF7t2NtOK0EQQ1H4/48+CAkdULjITy4jVp5LcmXvdk8mz08+UwSep7axzBMhY4eAEELGCIytoyGEjBEYW0dDCBkjMLaOhhAyRmBsHQ0hZIzA2DoaQsgYgbF1NISQMQJj62gIIWMExtbREELGCIytoyGEjBEYW0dDCBkjMLaOhhAyRmBsHQ0hZIzA2DoaQsgYgbF1NISQMQJj62gIIWMExtbREELGCIytoyGEjBEYW0dDCPlP4N8Yi8d1Tg7rSejbNyfkixNJyPc1PWFzEqohY6fgGyGXh+N1pccr9GSfk1BCNCT5QachD5Qu2+rKWrmzPxwKDdGQzzfp5TUxcSI15J0AIeNv6pdt9VD3UP/6l/jlqXRlubJ+fD+cOCBLDUneppszJ2xOQv3b+zv+y2qe/iTr5LCehGrIZkOSU/rnZi4b8udgJ1+YkIRScYaQIuwkipCEUnGGkCLsJIqQhFJxhpAi7CSKkIRScYaQIuwkipCEUnGGkCLsJIqQhFJxhpAi7CSKkIRScYaQIuwkipCEUnGGkCLsJIqQhFJxhpAi7CSKkIRScYaQIuwkipCEUnGGkCLsJIqQhFJxhpAi7CSKkIRScYaQIuwkipCEUnGGkCLsJIqQhFJxhpAi7CSKkIRScYaQIuwkipCEUnGGkCLsJIqQhFJxhpAi7CSKkIRScYaQIuwk6gUhYSBl3MCJmAAAAABJRU5ErkJggg=="`;

exports[`[Rectangle] should draw square on bottom right corner 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAAAXNSR0IArs4c6QAAAYdJREFUeF7t2DEOg0AQxVC4/6GhRNCtZgsLvfSMHJuvSDkPn5SBM0UD5hAk9hIIIkjMQAzHQgSJGYjhWIggMQMxHAsRJGYghmMhgsQMxHAsRJCYgRiOhQgSMxDDsRBBYgZiOBYiSMxADMdCBIkZiOFYiCAxAzEcCxEkZiCGYyGCxAzEcCxEkJiBGI6FCBIzEMOxEEFiBmI4FiJIzEAMx0IEiRmI4ViIIDEDMRwLESRmIIZjIYLEDMRwLESQmIEYjoUIEjMQw7EQQWIGYjgWIkjMQAzHQgSJGYjhWIggMQMxHAsRJGYghmMhgsQMxHAsRJCYgRiOhQgSMxDDsRBBYgZiOBYiSMxADMdCBIkZiOFYiCAxAzEcCxEkZiCGYyGCxAzEcCxEkJiBGI6FCBIzMMe55ieeCxYytynI3OHWC4Js1Tk/Jsjc4dYL3yCrPwOv51cf3vpNfnJMkFhIQQSJGYjhWIggMQMxHAsRJGYghmMhgsQMxHAsRJCYgRiOPxcFiRmI4VjIn4PcQ6sgUZ8oVqIAAAAASUVORK5CYII="`;

exports[`[Rectangle] should draw square on top left corner 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAAAXNSR0IArs4c6QAAAalJREFUeF7t2EEKwgAUxND2/odWupVuCgNGfO79hsRR6Hkcx+vYvs7tuf+6dskTJNRckFCMC0WQHwjy9D/g8yfv6ftjSr6Lc7eQp0IFGTYUZChzcUqQhcXhDUGGMhenBFlYHN4QZChzcUqQhcXhDUGGMhenBFlYHN4QZChzcUqQhcXhDQ8XhzIXpwRZWBzeEGQoc3FKkIXF4Y2nj9qHH+3UnQFBYt8LQQSJGYjhWIggMQMxHAsRJGYghmMhgsQMxHAsRJCYgRiOhQgSMxDDsRBBYgZiOBYiSMxADMdCBIkZiOFYiCAxAzEcCxEkZiCGYyGCxAzEcCxEkJiBGI6FCBIzEMOxEEFiBmI4FiJIzEAMx0IEiRmI4ViIIDEDMRwLESRmIIZjIYLEDMRwLESQmIEYjoUIEjMQw7EQQWIGYjgWIkjMQAzHQgSJGYjhWIggMQMxHAsRJGYghmMhgsQMxHAsRJCYgRiOhQgSMxDDsRBBYgZiOBYiSMxADMdCBIkZiOFYiCAxAzEcCxEkZiCGYyGCxAzEcCxEkJiBGI6FCBIzEMOxEEFiBmI4b1p5IGUONKIcAAAAAElFTkSuQmCC"`;
